File: VisualStudio_IDE/MmgGameApiCs/MmgGameApiCs/src/net/middlemind/DungeonTrap/ChapterE5_Phase3_CompleteGame/MdtUiPoints.cs
using System;
using net.middlemind.MmgGameApiCs.MmgBase;

namespace net.middlemind.DungeonTrap.ChapterE5_Phase3_CompleteGame
{
    /// <summary>
    /// A class that represents a UI element, the health bar.
    ///
    /// @author Victor G.Brusca, Middlemind Games
    /// 09/22/2020
    /// </summary>
    public class MdtUiPoints : MdtBase
    {
        /// <summary>
        /// The subject of this UI element.
        /// </summary>
        public MmgBmp subj = null;

        /// <summary>
        /// A bool flag indicating if the associated UI points are done displaying.
        /// </summary>
        public bool isFinished = false;

        /// <summary>
        /// A private variable used in internal methods.
        /// </summary>
        private bool lret = false;

        /// <summary>
        /// The current time this object has spent floating.
        /// </summary>
        public long floatTiming = 0;

        /// <summary>
        /// The total time this object can spend floating.
        /// </summary>
        public long floatTimingTotal = 500;

        /// <summary>
        /// The game screen this MdtChar belongs to.
        /// </summary>
        public ScreenGame screen = null;

        /// <summary>
        /// The player that triggered this UI object.
        /// </summary>
        public MdtPlayerType player;

        /// <summary>
        /// A base constructor that takes no arguments and loads object resources automatically.
        /// </summary>
        /// <param name="Subj">The subject image for this UI health bar.</param>
        /// <param name="Player">The player type this UI health bar is associated with.</param>
        /// <param name="Screen">The ScreenGame object this health bar belongs to.</param>
        /// <param name="StartPosition">The start position of the UI points object.</param>
        public MdtUiPoints(MmgBmp Subj, MdtPlayerType Player, ScreenGame Screen, MmgVector2 StartPosition)
        {
            SetSubj(Subj);
            SetMdtType(MdtObjType.UI);
            SetMdtSubType(MdtObjSubType.UI_POINTS);
            SetWidth(subj.GetWidth());
            SetHeight(subj.GetHeight());
            SetScreen(Screen);
            SetPosition(StartPosition);
        }

        /// <summary>
        /// A constructor that allows you to specify the subject of the object.
        /// </summary>
        /// <param name="Subj">The subject of the object.</param>
        public MdtUiPoints(MmgBmp Subj)
        {
            SetSubj(Subj);
            SetMdtType(MdtObjType.UI);
            SetMdtSubType(MdtObjSubType.UI_POINTS);
            SetWidth(subj.GetWidth());
            SetHeight(subj.GetHeight());
        }

        /// <summary>
        /// Gets the player that triggered this UI object.
        /// </summary>
        /// <returns>The player that triggered this UI object.</returns>
        public MdtPlayerType GetPlayer()
        {
            return player;
        }

        /// <summary>
        /// Sets the player that triggered this UI object.
        /// </summary>
        /// <param name="p">The player that triggered this UI object.</param>
        public void SetPlayer(MdtPlayerType p)
        {
            player = p;
        }

        /// <summary>
        /// Gets the game screen this character belongs to.
        /// </summary>
        /// <returns>The game screen this character belongs to.</returns>
        public ScreenGame GetScreen()
        {
            return screen;
        }

        /// <summary>
        /// Sets the game screen this character belongs to.
        /// </summary>
        /// <param name="o">The game screen this character belongs to.</param>
        public void SetScreen(ScreenGame o)
        {
            screen = o;
        }

        /// <summary>
        /// Gets the subject of the object.
        /// </summary>
        /// <returns>The subject of the object.</returns>
        public MmgBmp GetSubj()
        {
            return subj;
        }

        /// <summary>
        /// Sets the subject of the object.
        /// </summary>
        /// <param name="Subj">The subject of the object. </param>
        public void SetSubj(MmgBmp Subj)
        {
            subj = Subj;
        }

        /// <summary>
        /// Gets a bool indicating if this object is done displaying.
        /// </summary>
        /// <returns>A bool indicating if this object is done displaying.</returns>
        public bool GetIsFinished()
        {
            return isFinished;
        }

        /// <summary>
        /// Sets a bool indicating if this object is done displaying.
        /// </summary>
        /// <param name="b">A bool indicating if this object is done displaying.</param>
        public void SetIsFinished(bool b)
        {
            isFinished = b;
        }

        /// <summary>
        /// Gets the time this object has spent floating.
        /// </summary>
        /// <returns>The time this object has spent floating.</returns>
        public long GetFloatTiming()
        {
            return floatTiming;
        }

        /// <summary>
        /// Sets the time this object has spent floating.
        /// </summary>
        /// <param name="f">The time this object has spent floating.</param>
        public void SetFloatTiming(long f)
        {
            floatTiming = f;
        }

        /// <summary>
        /// Gets the total time this object can spend floating.
        /// </summary>
        /// <returns>The total time this object can spend floating.</returns>
        public long GetFloatTimingTotal()
        {
            return floatTimingTotal;
        }

        /// <summary>
        /// Sets the total time this object can spend floating.
        /// </summary>
        /// <param name="f">The total time this object can spend floating.</param>
        public void SetFloatTimingTotal(long f)
        {
            floatTimingTotal = f;
        }

        /// <summary>
        /// Sets the position of the object.
        /// </summary>
        /// <param name="v">The position of the object.</param>
        public override void SetPosition(MmgVector2 v)
        {
            base.SetPosition(v);
            subj.SetPosition(v);
        }

        /// <summary>
        /// Sets the position of the object.
        /// </summary>
        /// <param name="x">The X coordinate of the object.</param>
        /// <param name="y">The Y coordinate of the object.</param>
        public override void SetPosition(int x, int y)
        {
            base.SetPosition(x, y);
            subj.SetPosition(x, y);
        }

        /// <summary>
        /// Sets the X coordinate of the object.
        /// </summary>
        /// <param name="i">The X coordinate of the object.</param>
        public override void SetX(int i)
        {
            base.SetX(i);
            subj.SetX(i);
        }

        /// <summary>
        /// Sets the Y coordinate of the object.
        /// </summary>
        /// <param name="i">The Y coordinate of the object.</param>
        public override void SetY(int i)
        {
            base.SetY(i);
            subj.SetY(i);
        }

        /// <summary>
        /// The MmgUpdate method used to call the update method of the child objects.
        /// </summary>
        /// <param name="updateTick">The update tick number. </param>
        /// <param name="currentTimeMs">The current time in the game in milliseconds.</param>
        /// <param name="msSinceLastFrame">The number of milliseconds between the last frame and this frame.</param>
        /// <returns>A bool indicating if any work was done this game frame.</returns>
        public override bool MmgUpdate(int updateTick, long currentTimeMs, long msSinceLastFrame)
        {
            lret = false;
            if (isVisible == true)
            {
                subj.MmgUpdate(updateTick, currentTimeMs, msSinceLastFrame);

                floatTiming += msSinceLastFrame;
                if (floatTiming >= floatTimingTotal)
                {
                    SetIsVisible(false);
                    SetIsFinished(true);
                    screen.RemoveObj(this);
                }
                else
                {
                    SetY(GetY() - MmgHelper.ScaleValue(3));
                }

                lret = true;
            }
            return lret;
        }

        /// <summary>
        /// Base draw method, handles drawing this class.
        /// </summary>
        /// <param name="p">The MmgPen used to draw this object.</param>
        public override void MmgDraw(MmgPen p)
        {
            if (isVisible == true)
            {
                subj.MmgDraw(p);
            }
        }
    }
}

File: NetBeans_IDE/MmgGameApiJava/cfg/asset_src/dungeon_trap_other_classes/MdtUiPoints.java
package game.jam.DungeonTrap;

import net.middlemind.MmgGameApiJava.MmgBase.MmgBmp;
import net.middlemind.MmgGameApiJava.MmgBase.MmgHelper;
import net.middlemind.MmgGameApiJava.MmgBase.MmgPen;
import net.middlemind.MmgGameApiJava.MmgBase.MmgVector2;

/**
 * A class that represents a UI element, the health bar.
 * 
 * @author Victor G. Brusca, Middlemind Games
 * 09/22/2020
 */
public class MdtUiPoints extends MdtBase {

    /**
     * The subject of this UI element.
     */
    public MmgBmp subj = null;
    
    /**
     * A boolean flag indicating if the associated UI points are done displaying.
     */
    public boolean isFinished = false;
        
    /**
     * A private variable used in internal methods.
     */
    private boolean lret = false;    
    
    /**
     * The current time this object has spent floating.
     */
    public long floatTiming = 0;
    
    /**
     * The total time this object can spend floating.
     */
    public long floatTimingTotal = 500;

    /**
     * The game screen this MdtChar belongs to.
     */
    public ScreenGame screen = null;
        
    /**
     * The player that triggered this UI object.
     */
    public MdtPlayerType player;
    
    /**
     * A base constructor that takes no arguments and loads object resources automatically.
     */
    public MdtUiPoints(MmgBmp Subj, MdtPlayerType Player, ScreenGame Screen, MmgVector2 StartPosition) {
        SetSubj(Subj);
        SetMdtType(MdtObjType.UI);
        SetMdtSubType(MdtObjSubType.UI_POINTS);
        SetWidth(subj.GetWidth());
        SetHeight(subj.GetHeight());
        SetScreen(Screen);
        SetPosition(StartPosition);
    }

    /**
     * A constructor that allows you to specify the subject of the object.
     * 
     * @param Subj      The subject of the object.
     */
    public MdtUiPoints(MmgBmp Subj) {
        SetSubj(Subj);
        SetMdtType(MdtObjType.UI);
        SetMdtSubType(MdtObjSubType.UI_POINTS);
        SetWidth(subj.GetWidth());
        SetHeight(subj.GetHeight());        
    }    
    
    
    /**
     * Gets the player that triggered this UI object.
     * 
     * @return      The player that triggered this UI object. 
     */
    public MdtPlayerType GetPlayer() {
        return player;
    }

    /**
     * Sets the player that triggered this UI object.
     * 
     * @param p     The player that triggered this UI object. 
     */
    public void SetPlayer(MdtPlayerType p) {
        player = p;
    }
        
    /**
     * Gets the game screen this character belongs to.
     * 
     * @return      The game screen this character belongs to. 
     */
    public ScreenGame GetScreen() {
        return screen;
    }

    /**
     * Sets the game screen this character belongs to.
     * 
     * @param o     The game screen this character belongs to. 
     */
    public void SetScreen(ScreenGame o) {
        screen = o;
    }    
    
    /**
     * Gets the subject of the object.
     * 
     * @return      The subject of the object.
     */
    public MmgBmp GetSubj() {
        return subj;
    }

    /**
     * Sets the subject of the object.
     * 
     * @param Subj      The subject of the object. 
     */
    public void SetSubj(MmgBmp Subj) {
        subj = Subj;
    }

    /**
     * Gets a boolean indicating if this object is done displaying.
     * 
     * @return      A boolean indicating if this object is done displaying. 
     */
    public boolean GetIsFinished() {
        return isFinished;
    }

    /**
     * Sets a boolean indicating if this object is done displaying.
     * 
     * @param b     A boolean indicating if this object is done displaying. 
     */
    public void SetIsFinished(boolean b) {
        isFinished = b;
    }

    /**
     * Gets the time this object has spent floating.
     * 
     * @return      The time this object has spent floating.
     */
    public long GetFloatTiming() {
        return floatTiming;
    }

    /**
     * Sets the time this object has spent floating.
     * 
     * @param f     The time this object has spent floating. 
     */
    public void SetFloatTiming(long f) {
        floatTiming = f;
    }

    /**
     * Gets the total time this object can spend floating.
     * 
     * @return      The total time this object can spend floating.
     */
    public long GetFloatTimingTotal() {
        return floatTimingTotal;
    }

    /**
     * Sets the total time this object can spend floating.
     * 
     * @param f     The total time this object can spend floating. 
     */
    public void SetFloatTimingTotal(long f) {
        floatTimingTotal = f;
    }
    
    /**
     * Sets the position of the object.
     * 
     * @param v     The position of the object.
     */
    @Override
    public void SetPosition(MmgVector2 v) {
        super.SetPosition(v);
        subj.SetPosition(v);
    }
    
    /**
     * Sets the position of the object.
     * 
     * @param x     The X coordinate of the object.
     * @param y     The Y coordinate of the object.
     */
    @Override
    public void SetPosition(int x, int y) {
        super.SetPosition(x, y);
        subj.SetPosition(x, y);
    }
    
    /**
     * Sets the X coordinate of the object.
     * 
     * @param i     The X coordinate of the object.
     */
    @Override
    public void SetX(int i) {
        super.SetX(i);
        subj.SetX(i);
    }
    
    /**
     * Sets the Y coordinate of the object.
     * 
     * @param i     The Y coordinate of the object. 
     */
    @Override
    public void SetY(int i) {
        super.SetY(i);
        subj.SetY(i);
    }       
   
    /**
     * The MmgUpdate method used to call the update method of the child objects.
     * 
     * @param updateTick            The update tick number. 
     * @param currentTimeMs         The current time in the game in milliseconds.
     * @param msSinceLastFrame      The number of milliseconds between the last frame and this frame.
     * @return                      A boolean indicating if any work was done this game frame.
     */
    @Override
    public boolean MmgUpdate(int updateTick, long currentTimeMs, long msSinceLastFrame) {
        lret = false;
        if (isVisible == true) {
            subj.MmgUpdate(updateTick, currentTimeMs, msSinceLastFrame);
            
            floatTiming += msSinceLastFrame;
            if(floatTiming >= floatTimingTotal) {
                SetIsVisible(false);
                SetIsFinished(true);  
                screen.RemoveObj(this);
            } else {
                SetY(GetY() - MmgHelper.ScaleValue(3));
            }
            
            lret = true;
        }
        return lret;
    }
        
    /**
     * Base draw method, handles drawing this class.
     *
     * @param p     The MmgPen used to draw this object.
     */
    @Override
    public void MmgDraw(MmgPen p) {
        if (isVisible == true) {
            subj.MmgDraw(p);
        }
    }    
}